### ISSUE TITLE
Subclassing networkx with Mixin class defined with Python Protocol results in an error during graph creation

### ISSUE THREAD

**Issue by adam2392** [networkx/networkx#6243]
### Current Behavior

If I define a subclass of a networkx graph and augment it with some mixin class that defines some extra functions everything works as expected. However, if I subclass a Python "Protocol" class in my mixin class to allow mypy to do its thing and for me to define an API interface, there is an error.

### Expected Behavior

`nx.to_networkx_graph(ts_edges, create_using=Graph)` should work as expected. 

The error arises due to the check inside `empty_graph ` for `elif type(create_using) is type:`.

```
def empty_graph(n=0, create_using=None, default=Graph):
    if create_using is None:
        G = default()
    elif type(create_using) is type:
        G = create_using()
    elif not hasattr(create_using, "adj"):
        raise TypeError("create_using is not a valid NetworkX graph type or instance")
    else:
        # create_using is a NetworkX style Graph
        create_using.clear()
        G = create_using
```

### Steps to Reproduce

The following code results in an error:

```
    from typing import Protocol

    import networkx as nx

    class Mixin(Protocol):
        _adj: Protocol

        def new_method(self):
            pass

    class BaseGraph(nx.DiGraph):
        def __init__(self, incoming_graph_data=None, **attr):
            super().__init__(incoming_graph_data, **attr)

    class Graph(Mixin, BaseGraph):
        def __init__(self, incoming_graph_data=None, **attr):
            super().__init__(incoming_graph_data, **attr)

    ts_edges = [
        (("x1", -1), ("x1", 0)),
        (("x1", -1), ("x2", 0)),
        (("x3", -1), ("x2", 0)),
        (("x3", -1), ("x3", 0)),
    ]
    G = Graph(ts_edges)
    new_G = Graph(G)

    G = nx.to_networkx_graph(ts_edges, create_using=Graph)
```

### Environment

Python version: 3.9
NetworkX version: 3.0b1.dev0

**Comment by dschult** [comment]
That's funny...  The typing.Protocol class changes the return value of the type() function, breaking our type-checking... :}  Perhaps we need to go back to ducktyping where we just try to create a graph with `create_using` and if it doesn't break we assume that all is OK.

OK...  more seriously, What does the `typing.Protocol` say about how to check the type of the subclass? In particular, we want to check if `create_using` is a `type` class that we could use to build an instance.

**Comment by boothby** [comment]
The `inspect` module has an official `isclass` method for this purpose.

**Comment by dschult** [comment]
That seems like the right fix: In empty_graph, replace the `elif type(create_using) is type:` with
```python
   elif isinstance(create_using, type):
```

I guess the workaround in the meantime is to make sure `create_using` is an instance by adding () to get:

```nx.to_networkx_graph(ts_edges, create_using=Graph())```

**Comment by adam2392** [comment]
Want me to start a PR?

**Comment by dschult** [comment]
Thanks for the offer -- I guess I should have waited... :} 
Anyway, can you see if #6244 does what you want? I added a test that broke before and this fixes it, but not sure it is the best test...

**Comment by adam2392** [comment]
Yes so this gets rid of the error due to the type checking. But another issue arised. 

Some more specifics are that I am trying to leverage networkx 's generality to store essentially anything "hashable" as a node. Therefore, I can store a 2-tuple as nodes to create a time-series graph, with the first index being the variable name and the second index being the time-lag. I then override the `add_edge` function to do an additional check:

```
def add_edge(...):
       ....
        # this checks the added nodes against a graph parameter
        if node[1] < -self.graph['max_lag']:
            raise ValueError(f"Lag {node[1]} cannot be greater than set max_lag {self.max_lag}.")
```

So I guess the deeper problem arises also from the fact that inside `nx.Graph.__init__`, there is this graph attribute updating AFTER adding the edge data.

```
        if incoming_graph_data is not None:
            convert.to_networkx_graph(incoming_graph_data, create_using=self)
        # load graph attributes (must be after convert)
        self.graph.update(attr)
```

In summary, I am trying to build a graph that has a global graph property (i.e. max-lag) that constrains what type of nodes are able to be added into the graph. However, if I store this into the `graph` attribute dictionary, then adding edges upon instantiation of the class is not allowed:

```
G = graph_func(edges, max_lag=4)  # error
G = graph_func(max_lag=4)
G.add_edges_from(edges)  # no error
```

**Comment by dschult** [comment]
Why not check the edges before sending them into the graph object? Something like:
```python
ok_edges = (e for e in edges if (e[0][1] < max_lag and e[1][1] < max_lag))
G = graph_func(ok_edges, max_lag=4)
```

**Comment by adam2392** [comment]
This gets around the instantiation part but not when you want to add edges after instantiation. 

Then you would need to do the checks every time you call "add_edge" outside the function call itself?

**Comment by dschult** [comment]
You can still do the checks each `G.add_edge`...  
And, of course, you can overwrite the `__init__` function to make max_lag available to the add_edges_from method before G.graph is created.  Lots of options... when you are the creator...

**Comment by adam2392** [comment]
I was going to store max lag in the graph attributes. Why do graph attributes need to be instantiated after edge creation?

**Comment by dschult** [comment]
They don't....  And I would have said the same thing about "why does edge creation need to be after graph attribute loading?".  It doesn't...  Someone might need to look at the graph attribute in order to set the edges. And someone might need to look at edges in order to set the graph attribute...  But almost always, there isn't any interaction between the inputs to the Graph constructor. 

Perhaps there maybe shouldn't be any inputs allowed during instantiation. Just create the graph and then add the features you want. But we liked the idea of allowing lots of convenient features.  Perhaps too many...  As you are finding... they may get in each other's way. But we don't have all the possible features that we could -- for example, nodes can't be specified in the "data".  Basically, we put as many features as we could fit, and we're relying on the users to police themselves.  

If someone (like you) wants to do fancy stuff, they can do fancy stuff... Change the order of operations for your subclass if that helps your subclass. :}

**Comment by adam2392** [comment]
Ah fair point. Sounds good!

In the meantime, https://github.com/networkx/networkx/pull/6244 seems like it solves the issue at hand.
